contracts/EthlanceJobs.sol

```diff
@@ -0,0 +1,851 @@
+      IERC721(jobs[_jobId].token).transferFrom(_sender,
+                                               address(this),
+                                               _amount);
+    } else {
+      revert();
+    }
+
+    emit ContributionAdded(_jobId,
+                           jobs[_jobId].contributions.length - 1, // The new contributionId
+                           _sender,
+                           _amount);
+  }
+
+  /// @dev refundContribution(): Allows users to refund the contributions they've
+  ///                            made to a particular job, but only if the job
+  ///                            has not yet paid out
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _contributionId the index of the contribution being refunded
+  function refundContribution(
+                              address _sender,
+                              uint _jobId,
+                              uint _contributionId)
+    public
+    senderIsValid(_sender)
+    validateJobArrayIndex(_jobId)
+    validateContributionArrayIndex(_jobId, _contributionId)
+    onlyContributor(_sender, _jobId, _contributionId)
+    hasNotPaid(_jobId)
+    hasNotRefunded(_jobId, _contributionId)
+    callNotStarted
+  {
+
+    Contribution storage contribution = jobs[_jobId].contributions[_contributionId];
+
+    contribution.refunded = true;
+
+    transferTokens(_jobId, contribution.contributor, contribution.amount); // Performs the disbursal of tokens to the contributor
+
+    emit ContributionRefunded(_jobId, _contributionId);
+  }
+
+  /// @dev refundMyContributions(): Allows users to refund their contributions in bulk
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _contributionIds the array of indexes of the contributions being refunded
+  function refundMyContributions(
+                                 address _sender,
+                                 uint _jobId,
+                                 uint[] memory _contributionIds)
+    public
+    senderIsValid(_sender)
+  {
+    for (uint i = 0; i < _contributionIds.length; i++){
+      refundContribution(_sender, _jobId, _contributionIds[i]);
+    }
+  }
+
+  /// @dev refundContributions(): Allows users to refund their contributions in bulk
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _issuerId the index of the issuer who is making the call
+  /// @param _contributionIds the array of indexes of the contributions being refunded
+  function refundContributions(
```

Comment:
Need a check to make sure the user owns the given contribution IDs

Reply:
I don't think it is needed since contribution amounts and contributors are in sync.